ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,399 @@
+            Iterable::Slice(ident, range) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                build_slice_ident_expression(ident, ident_type, range.start(), i)
+            }
+        },
+    }
+}
+
+/// Parses a named trace access in a list comprehension expression.
+///
+/// # Errors
+/// - Returns an error if the iterable is an identifier and that identifier does not correspond to
+///   a trace column.
+/// - Returns an error if the iterable is a range.
+/// - Returns an error if the iterable is a slice and that identifier does not correspond to a
+///   trace column.
+fn parse_named_trace_access(
+    named_trace_access: &NamedTraceAccess,
+    iterable_context: &IterableContext,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    let iterable = iterable_context.get(&Identifier(named_trace_access.name().to_string()));
+    match iterable {
+        // if the corresponding iterable is not present in the iterable context that means the
+        // trace column is not part of the list comprehension and we just return it as it is.
+        None => Ok(Expression::NamedTraceAccess(named_trace_access.clone())),
+        Some(iterable_type) => match iterable_type {
+            Iterable::Identifier(ident) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                match ident_type {
+                    IdentifierType::TraceColumns(size) => {
+                        validate_access(i, size.size())?;
+                        Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                            ident.clone(),
+                            i,
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(format!(
+                        "Iterable {ident} should be trace columns"
+                    )))?,
+                }
+            }
+            Iterable::Range(_) => Err(SemanticError::InvalidListComprehension(format!(
+                "Iterable cannot be of type Range for named trace access {}",
+                named_trace_access.name()
+            ))),
+            Iterable::Slice(ident, range) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                match ident_type {
+                    IdentifierType::TraceColumns(trace_columns) => {
+                        validate_access(i, trace_columns.size())?;
+                        Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                            ident.clone(),
+                            range.start() + i,
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(format!(
+                        "Iterable {ident} should be trace columns"
+                    )))?,
```

Comment:
nit: this grammar doesn't sound quite right. Should this be "Iterable {ident} should be a trace column" or "Iterable {ident} should contain trace columns"?